File: electron-markdown-converter/src/lib/converter.ts
import * as cheerio from 'cheerio';
import { JSDOM } from 'jsdom';
import { convert } from 'html-to-text';

export interface ConversionOptions {
  selector?: string;
  bodyWidth?: number;
  preserveLinks?: boolean;
  preserveImages?: boolean;
  preserveEmphasis?: boolean;
  preserveTables?: boolean;
}

export async function convertHtmlToMarkdown(
  htmlContent: string,
  options: ConversionOptions = {}
): Promise<string> {
  try {
    // Set default options
    const {
      selector,
      bodyWidth = 0,
      preserveLinks = true,
      preserveImages = true,
      preserveEmphasis = true,
      preserveTables = true
    } = options;

    // Configure conversion options
    const conversionOptions = {
      wordwrap: bodyWidth,
      selectors: [
        { selector: 'a', options: { ignoreHref: !preserveLinks } },
        { selector: 'img', options: { ignoreImage: !preserveImages } },
        { selector: 'table', options: { skipTables: !preserveTables } },
        { selector: 'em,strong,b,i', options: { ignoreEmphasis: !preserveEmphasis } }
      ]
    };

    let contentToConvert = htmlContent;

    if (selector) {
      // Use cheerio for CSS selector-based content extraction
      const $ = cheerio.load(htmlContent);
      const selectedContent = $(selector);
      
      if (selectedContent.length > 0) {
        contentToConvert = selectedContent.map((_, el) => $.html(el)).get().join('\n');
      } else {
        // Fallback to basic DOM extraction if selector doesn't match
        const dom = new JSDOM(htmlContent);
        const { document } = dom.window;
        contentToConvert = document.body.innerHTML;
      }
    }

    // Convert to markdown
    const markdown = convert(contentToConvert, conversionOptions);
    return markdown.trim();
  } catch (error) {
    throw new Error(`Conversion error: ${error.message}`);
  }
}

File: converter.py
from bs4 import BeautifulSoup
import html2text
import trafilatura

def convert_html_to_markdown(html_content, selector=None):
    """
    Convert HTML content to Markdown format
    """
    try:
        # Initialize HTML to text converter
        converter = html2text.HTML2Text()
        converter.body_width = 0  # Disable line wrapping
        converter.ignore_links = False
        converter.ignore_images = False
        converter.ignore_emphasis = False
        converter.ignore_tables = False

        if selector:
            # Parse HTML and extract content based on selector
            soup = BeautifulSoup(html_content, 'html.parser')
            selected_content = soup.select(selector)
            if selected_content:
                html_content = ''.join(str(element) for element in selected_content)
            else:
                # If selector doesn't match, use trafilatura's extraction
                html_content = trafilatura.extract(html_content, include_tables=True, 
                                                 include_links=True, include_images=True,
                                                 output_format='html')

        # Convert to markdown
        markdown = converter.handle(html_content)
        return markdown.strip()
    except Exception as e:
        raise Exception(f"Conversion error: {str(e)}")
